app/frontend/src/features/communities/CommunityPage/JoinCommunityButton.tsx

```diff
@@ -0,0 +1,43 @@
+import Button from "components/Button";
+import Snackbar from "components/Snackbar";
+import {
+  JOIN_COMMUNITY,
+  LEAVE_COMMUNITY,
+} from "features/communities/constants";
+import { Community } from "pb/communities_pb";
+import { communityKey } from "queryKeys";
+import { useMutation, useQueryClient } from "react-query";
+import { service } from "service";
+
+export default function JoinCommunityButton({
+  community,
+}: {
+  community: Community.AsObject;
+}) {
+  const queryClient = useQueryClient();
+  const join = useMutation<void, Error>(async () => {
+    await service.communities.joinCommunity(community.communityId);
+    await queryClient.refetchQueries(communityKey(community.communityId));
```

Comment:
I tried that first, but it's the same old problem where the button stops loading and still displays the old state while it's invalidating. Is there a better way to avoid that than this?

Reply:
Ohh because by the time line 19 is done, `isLoading` is false? Can we get the query's refetching/loading state out from the query client? Otherwise can't think of anything better without introducing another spinner like I did in other places, which kinda feels a bit suboptimal imo now I think more about this :/

Reply:
This looks promising - https://react-query.tanstack.com/reference/useIsFetching

Then maybe we can combine that on `isLoading` with another `||` ?

Reply:
How about I do an optimistic update in `onSuccess`

Reply:
How about I do an optimistic update in `onSuccess`

Reply:
Hmm yeah that'd also work, since this isn't an infinite query, it's a lot easier to optimistic update and so seems reasonable

Reply:
Maybe that `isFetching` thing is a good workaround for infinite queries though, good to have looked that up :D

Reply:
It might appear every 5 seconds though for the ping! 😱